### PR TITLE
Add all remaining `prif` interfaces

### PR DESCRIPTION
**Pull request by ktras**
Add all remaining `prif` interfaces based on design doc revision 0.2

### CODE REVIEW

File: src/prif.f90
Comment (bonachea) [inline src/prif.f90:1]: Missing new `prif_critical_type`, which should probably be just a renamed `prif_lock_type`

**Review by bonachea — CHANGES_REQUESTED** [review]
Overall LGTM. 

I did NOT attempt to exhaustively compare all the subroutine signatures for spec conformance, although that's something that eventually needs to be audited, and ideally supported with test code that ensures calls compile as intended.

**Review by everythingfunctional — APPROVED** [review]
A did a brief once-over and nothing jumps out. LGTM

File: src/prif.f90
Comment (ktras) [reply src/prif.f90:1]: Done.

**Review by bonachea — APPROVED** [review]
Removing `pure` from functions internal to caffeine is probably overkill, but I have no specific objection.

The important part is to match the specified `prif_` interfaces at the top.

**Comment by ktras** [comment]
> Removing `pure` from functions internal to caffeine is probably overkill, but I have no specific objection.
> 
> The important part is to match the specified `prif_` interfaces at the top.

@bonachea, I agree. I only removed `pure` from functions that I had to based on the cascading affects of `pure` being removed from some of the `prif_` interfaces that called other `pure` functions and no longer can. Thanks for the review!

**Pull request merged**